core/src/test/scala/integration/kafka/api/TransactionsTest.scala

```diff
@@ -151,208 +165,182 @@ class TransactionsTest extends KafkaServerTestHarness {
 
+    try {
       producer1.beginTransaction()
-      producer1.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, "1", "1", willBeCommitted = false))
-      producer1.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic2, "3", "3", willBeCommitted = false))
-      producer1.abortTransaction()
-
-      producer2.initTransactions()  // ok, will abort the open transaction.
-      producer2.beginTransaction()
-      producer2.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, "2", "4", willBeCommitted = true)).get()
-      producer2.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic2, "2", "4", willBeCommitted = true)).get()
-
-      try {
-        producer1.beginTransaction()
-        val result =  producer1.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, "1", "5", willBeCommitted = false))
-        val recordMetadata = result.get()
-        error(s"Missed a producer fenced exception when writing to ${recordMetadata.topic()}-${recordMetadata.partition()}. Grab the logs!!")
-        servers.foreach { case (server) =>
-          error(s"log dirs: ${server.logManager.logDirs.map(_.getAbsolutePath).head}")
-        }
-        fail("Should not be able to send messages from a fenced producer.")
-      } catch {
-        case e : ProducerFencedException =>
-          producer1.close()
-        case e : ExecutionException =>
-          assertTrue(e.getCause.isInstanceOf[ProducerFencedException])
-        case e : Exception =>
-          fail("Got an unexpected exception from a fenced producer.", e)
+      val result =  producer1.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, "1", "5", willBeCommitted = false))
+      val recordMetadata = result.get()
+      error(s"Missed a producer fenced exception when writing to ${recordMetadata.topic()}-${recordMetadata.partition()}. Grab the logs!!")
+      servers.foreach { case (server) =>
+        error(s"log dirs: ${server.logManager.logDirs.map(_.getAbsolutePath).head}")
       }
+      fail("Should not be able to send messages from a fenced producer.")
+    } catch {
+      case e : ProducerFencedException =>
+        producer1.close()
+      case e : ExecutionException =>
+        assertTrue(e.getCause.isInstanceOf[ProducerFencedException])
+      case e : Exception =>
+        fail("Got an unexpected exception from a fenced producer.", e)
+    }
 
-      producer2.commitTransaction()  // ok
+    producer2.commitTransaction()  // ok
 
-      val records = pollUntilExactlyNumRecords(consumer, 2)
-      records.zipWithIndex.foreach { case (record, i) =>
-        TestUtils.assertCommittedAndGetValue(record)
-      }
-    } finally {
-      consumer.close()
-      producer1.close()
-      producer2.close()
+    val records = pollUntilExactlyNumRecords(consumer, 2)
+    records.zipWithIndex.foreach { case (record, i) =>
```

Comment:
That's actually from the original code. I copied the pattern from another test (can't remember which). Incidentally, I could not get the `zip` to work at the time, though I think I would if I tried now. (It's amazing how a couple of weeks of intensive scala stimulates the imagination).